src/many-protocol/src/lib.rs

```diff
@@ -1,616 +1,129 @@
+use coset::CoseSign1;
+use coset::CoseSign1Builder;
+use many_error::ManyError;
+use many_identity::{Address, Identity, Verifier};
+
 pub mod request;
 pub mod response;
 
-use std::collections::BTreeMap;
-
-use coset::cbor::value::Value;
-use coset::iana::Algorithm;
-use coset::CborSerializable;
-use coset::CoseKeySet;
-use coset::CoseSign1;
-use coset::CoseSign1Builder;
-use coset::HeaderBuilder;
-use coset::Label;
-pub use many_error::ManyError;
-use many_identity::cose_helpers::public_key;
-use many_identity::{Address, CoseKeyIdentity, CoseKeyIdentitySignature};
-use minicbor::Decode;
-pub use request::RequestMessage;
-pub use request::RequestMessageBuilder;
-pub use response::ResponseMessage;
-pub use response::ResponseMessageBuilder;
-use serde::{Deserialize, Serialize};
-use sha2::Digest;
-use signature::{Signature, Signer, Verifier};
+pub use request::{RequestMessage, RequestMessageBuilder};
+pub use response::{ResponseMessage, ResponseMessageBuilder};
 
 pub type ManyUrl = reqwest::Url;
 
-pub fn decode_request_from_cose_sign1(
-    sign1: CoseSign1,
-    allowed_origins: Option<Vec<ManyUrl>>,
+pub trait IdentityResolver: Send {
+    fn resolve_request(
+        &self,
+        envelope: &CoseSign1,
+        request: &RequestMessage,
+    ) -> Result<Address, ManyError>;
+    fn resolve_response(
+        &self,
+        envelope: &CoseSign1,
+        response: &ResponseMessage,
+    ) -> Result<Address, ManyError>;
+}
+
+impl IdentityResolver for Box<dyn IdentityResolver> {
+    fn resolve_request(
+        &self,
+        envelope: &CoseSign1,
+        request: &RequestMessage,
+    ) -> Result<Address, ManyError> {
+        (&**self).resolve_request(envelope, request)
+    }
+
+    fn resolve_response(
+        &self,
+        envelope: &CoseSign1,
+        response: &ResponseMessage,
+    ) -> Result<Address, ManyError> {
+        (&**self).resolve_response(envelope, response)
+    }
+}
```

Comment:
I'll comment on it only once, but this also applies to `many-server`.

What does this do, and how is it used? 

I don't get it.

Reply:
I'll remove it, that's a bit premature.

Reply:
Actually, I'll keep it there. It is the bit of code that takes a message and return the origin address of the message.

This is (currently) the hash of the public key (or the `kid` too), but it could be different. It's also something that requires more logic than reading a single field because it should match the signature.

This was done by the Verifier before (or the code that was doing the verification), but there are cases that become very complex and shouldn't be left to the verifier itself (like delegation).

I just now noticed that this logic was always missing. I'll re-add it to make sure we don't introduce a bug.

Reply:
`verify_1` now returns the address that matched the verification. Additionaly resolution steps might be required.